[WIP][IR][Constants] Change the semantic of `ConstantPointerNull` to represent an actual `nullptr` instead of a zero-value pointer

shiltian opened the pull request:
## Motivation

The value of a `nullptr` is not always `0`. For example, on AMDGPU, the `nullptr` in address spaces 3 and 5 is `0xffffffff`. Currently, there is no target-independent way to get this information, making it difficult and error-prone to handle null pointers in target-agnostic code.

We do have `ConstantPointerNull`, but it might be a little confusing and misleading. It represents a pointer with an all-zero value rather than necessarily a real `nullptr`. Therefore, to represent a real `nullptr` in address space `N`, we need to use `addrspacecast ptr null to ptr addrspace(N)` and it can't be folded.

## Proposal

In this PR, we change the semantic of `ConstantPointerNull` to represent an actual `nullptr` instead of a zero-value pointer. Here is the detailed changes.

### IR

* `ptr addrspace(N) null` will represent the actual `nullptr` in address space `N`.
* `ptr addrspace(N) zeroinitializer` will represent a zero-value pointer in address space `N`.

### API

* `Constant::getNullValue` will return a _null_ value. It is same as the current semantics except for the `PointerType`, which will return a real `nullptr` pointer.
* `Constant::getZeroValue` will return a zero value constant. It is completely same as the current semantics. To represent a zero-value pointer, a `ConstantExpr` will be used (effectively `inttoptr i8 0 to ptr addrspace(N)`).
* Correspondingly, there will be both `Constant::isNullValue` and `Constant::isZeroValue`.

The RFC is https://discourse.llvm.org/t/rfc-introduce-sentinel-pointer-value-to-datalayout/85265. It is a little bit old and the title might look different, but everything eventually converges to this change. An early attempt can be found in https://github.com/llvm/llvm-project/pull/131557, which has many valuable discussion as well.

This PR is still WIP but any early feedback is welcome. I'll include as many necessary code changes as possible in this PR, but eventually this needs to be carefully split into multiple PRs, and I'll do it after the changes look good to every one.

arichardson commented on llvm/lib/IR/ConstantFold.cpp:
```diff
@@ -126,6 +126,16 @@ Constant *llvm::ConstantFoldCastInstruction(unsigned opc, Constant *V,
+  // TODO: For the following two cases, we can fold the constant to a constant
+  // according to the data layout.
+
+  // We can't fold inttoptr(0) to ConstantNullPointer.
```
```suggestion
  // We can't fold inttoptr(0) to ConstantNullPointer without DataLayout.
```

arichardson commented on llvm/lib/IR/ConstantFold.cpp:
```diff
@@ -126,6 +126,16 @@ Constant *llvm::ConstantFoldCastInstruction(unsigned opc, Constant *V,
+  // according to the data layout.
+
+  // We can't fold inttoptr(0) to ConstantNullPointer.
+  if (opc == Instruction::IntToPtr && V->isZeroValue())
```
Could you add these folds to the data-layout aware constant folder?

arichardson commented on llvm/lib/IR/DataLayout.cpp:
```diff
@@ -193,9 +193,11 @@ constexpr DataLayout::PrimitiveSpec DefaultVectorSpecs[] = {
 
 // Default pointer type specifications.
-constexpr DataLayout::PointerSpec DefaultPointerSpecs[] = {
+const DataLayout::PointerSpec DefaultPointerSpecs[] = {
```
We could avoid the need for a global ctor/dtor if we just stored a plain integer value? Since we only support 0/-1 we could just rely on sign-extension? But of course that makes the getter slower so probably best to ignore this suggestion.

shiltian replied:
I thought about that, but we might want to allow a target to specify that the `nullptr` is an arbitrary bit pattern to be more future proof such that LLVM will not try any folding regarding those casts involving `nullptr`. I added `'c'` as the flag to specify that.

shiltian commented on llvm/lib/IR/Constants.cpp:
```diff
@@ -369,7 +378,10 @@ bool Constant::containsConstantExpression() const {
+/// means a constant with value '0', but for pointer types, it represents a
+/// nullptr constant. A nullptr isn't always a zero-value pointer in certain
+/// address spaces on some targets.
 Constant *Constant::getNullValue(Type *Ty) {
```
Most of the places where a `ConstantPointerNull` is constructed is via `Constant::getNull`, so if we do `Constant::getNullPtrValue`, then that will require pretty much every place to be rewritten with a `if-else` based on their type, which I tried locally to some extent a couple of months ago, and it was not fun.

shiltian commented on llvm/test/CodeGen/AMDGPU/addrspacecast.gfx6.ll:
```diff
@@ -182,7 +182,7 @@ define ptr addrspace(3) @addrspacecast_flat_null_to_group() {
 ; CHECK:       ; %bb.0:
 ; CHECK-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
-; CHECK-NEXT:    v_mov_b32_e32 v0, -1
+; CHECK-NEXT:    v_mov_b32_e32 v0, 0
```
This change is wrong. Need to find the right place to do the right value.

arichardson commented on llvm/lib/Analysis/ConstantFolding.cpp:
```suggestion
          FoldedValue = ConstantInt::get(CE->getContext(), NullPtrValue + BaseOffset);
```
plus appropriate handling of zero-extension

shiltian commented on llvm/test/CodeGen/AMDGPU/addrspacecast.r600.ll:
```diff
@@ -42,7 +42,7 @@ define amdgpu_kernel void @addrspacecast_flat_null_to_local(ptr addrspace(1) %ou
 ; CHECK-NEXT:    ALU clause starting at 4:
 ; CHECK-NEXT:     MOV * T0.X, literal.x,
-; CHECK-NEXT:    -1(nan), 0(0.000000e+00)
+; CHECK-NEXT:    0(0.000000e+00), 0(0.000000e+00)
```
Does R600 even have multiple address spaces? It doesn't look like the case from the data layout string.

arichardson commented on llvm/test/Instrumentation/AddressSanitizer/asan-scalable-vector.ll:
```diff
@@ -7,13 +7,16 @@ define void @test() #1 {
-; CHECK-NEXT:    call void @__asan_report_store8(i64 0) #[[ATTR4:[0-9]+]]
+; CHECK-NEXT:    br i1 [[TMP1]], label %[[BB5:.*]], label %[[BB6:.*]]
+; CHECK:       [[BB5]]:
+; CHECK-NEXT:    call void @__asan_report_store8(i64 ptrtoint (ptr null to i64)) #[[ATTR4:[0-9]+]]
```
I feel like we should really be passing a pointer to these hooks rather than an integer value, but also am surprised this isn't using the datalayout aware constant folding.

arichardson commented on llvm/test/Transforms/InstCombine/ptrtoint-nullgep.ll:
```diff
@@ -16,8 +16,14 @@ declare void @use_i64(i64)
+; LLPARSER-NEXT:    ret i64 ptrtoint (ptr addrspace(1) null to i64)
+;
+; INSTSIMPLIFY-LABEL: define {{[^@]+}}@constant_fold_ptrtoint_gep_zero() {
+; INSTSIMPLIFY-NEXT:    ret i64 ptrtoint (ptr addrspace(1) null to i64)
```
Hmm, shouldn't we have the datalayout in instsimplify, does this mean we don't use the datalayout aware constant folding logic there?

arichardson reviewed (commented):
This generally looks good to me, just a few comments about the test diffs.

shiltian commented:
Still WIP. With more code changes, there are more test failures. Fixing them one by one. :-)

arichardson commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -1561,6 +1583,24 @@ Constant *llvm::ConstantFoldCastOperand(unsigned Opcode, Constant *C,
+
+    // If the input is a nullptr, we can fold it to the corresponding
+    // nullptr in the destination address space.
+    if (C->isNullValue())
```
Is this actually true? Are null values in one address space always null in all others? I imagine this is almost always true, but maybe you could have some where this conversion is not valid?

arichardson replied:
Sounds good

arichardson commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -842,7 +842,7 @@ void Verifier::visitGlobalVariable(const GlobalVariable &GV) {
     // cannot be constant.
     if (GV.hasCommonLinkage()) {
-      Check(GV.getInitializer()->isNullValue(),
+      Check(GV.getInitializer()->isZeroValue(),
```
I wonder if we should add an optional data layout parameter to this function so it can return true for null pointers in the common case?

arichardson reviewed (commented):
A few more small comments

arichardson commented on llvm/test/Assembler/ConstantExprFoldCast.ll:
```diff
@@ -5,8 +5,8 @@
-; CHECK: addrspacecast
-; CHECK: addrspacecast
+; CHECK: ptr addrspace(1) null
+; CHECK: ptr null
```
Unrelated to this PR but these checks are not great, can we regen this test with UTC and --check-globals and pre-commit that?

arichardson commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -1497,6 +1497,21 @@ Constant *llvm::ConstantFoldCastOperand(unsigned Opcode, Constant *C,
+              DestTy, NullPtrValue->zextOrTrunc(DestTy->getScalarSizeInBits()));
+        } else {
+          llvm_unreachable("invalid nullptr value");
+        }
```
```suggestion
        assert(NullPtrValue->isZero() || NullPtrValue->isAllOnes());
        return ConstantInt::get(
              DestTy, NullPtrValue->zextOrTrunc(DestTy->getScalarSizeInBits()));
```
I believe this is equivalent and a bit simpler.

shiltian commented on llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -4275,8 +4275,15 @@ static void emitGlobalConstantImpl(const DataLayout &DL, const Constant *CV,
+      // We fall back to the default behavior of emitting a zero value if we
+      // can't get the null pointer value from the data layout.
+      AP.OutStreamer->emitIntValue(0, Size);
+    }
```
That sounds like a good idea. I can do that but I guess we'd need an RFC.

shiltian committed:
[WIP][IR][Constants] Change the semantic of `ConstantPointerNull` to represent an actual `nullptr` instead of a zero-value pointer

The value of a `nullptr` is not always `0`. For example, on AMDGPU, the `nullptr` in address spaces 3 and 5 is `0xffffffff`. Currently, there is no target-independent way to get this information, making it difficult and error-prone to handle null pointers in target-agnostic code.

We do have `ConstantPointerNull`, but it might be a little confusing and misleading. It represents a pointer with an all-zero value rather than necessarily a real `nullptr`. Therefore, to represent a real `nullptr` in address space `N`, we need to use `addrspacecast ptr null to ptr addrspace(N)` and it can't be folded.

In this PR, we change the semantic of `ConstantPointerNull` to represent an actual `nullptr` instead of a zero-value pointer. Here is the detailed changes.

* `ptr addrspace(N) null` will represent the actual `nullptr` in address space `N`.
* `ptr addrspace(N) zeroinitializer` will represent a zero-value pointer in address space `N`.

* `Constant::getNullValue` will return a _null_ value. It is same as the current semantics except for the `PointerType`, which will return a real `nullptr` pointer.
* `Constant::getZeroValue` will return a zero value constant. It is completely same as the current semantics. To represent a zero-value pointer, a `ConstantExpr` will be used (effectively `inttoptr i8 0 to ptr addrspace(N)`).
* Correspondingly, there will be both `Constant::isNullValue` and `Constant::isZeroValue`.

The RFC is https://discourse.llvm.org/t/rfc-introduce-sentinel-pointer-value-to-datalayout/85265. It is a little bit old and the title might look different, but everything eventually converges to this change. An early attempt can be found in https://github.com/llvm/llvm-project/pull/131557, which has many valuable discussion as well.

This PR is still WIP but any early feedback is welcome. I'll include as many necessary code changes as possible in this PR, but eventually this needs to be carefully split into multiple PRs, and I'll do it after the changes look good to every one.

arsenm commented on llvm/include/llvm/IR/DataLayout.h:
```diff
@@ -84,6 +84,11 @@ class DataLayout {
+    /// all-zeros, all-ones, or std::nullopt. Since we don't have a way to
+    /// represent an arbitrary bit pattern, we use std::nullopt to represent the
+    /// case where the nullptr value is neither 0 nor -1.
+    std::optional<APInt> NullPtrValue;
```
"Don't have a way to represent arbitrary bit pattern", but this is stored as APInt.

shiltian replied:
It is because we can't assume the size of the pointer, even though they are `0U` or `~0U`. This function can just return a `std::optional<bool>`, where `true` means `~0U`, and `false` means `0U`, but is that a good design?

arsenm reviewed (commented):
Browser is dying on this

arichardson commented on llvm/test/Analysis/Delinearization/gcd_multiply_expr.ll:
```diff
@@ -20,7 +20,7 @@
 @g = common global i32 0, align 4
 @h = common global i32 0, align 4
-@f = common global ptr null, align 4
+@f = common global ptr zeroinitializer, align 4
```
Why do you need to update any non-AMDGPU tests here to use zeroinitializer? Is there some pass that gives different output since it does not have the datalayout available? I feel like it would be a lot easier to read this diff if the existing `null` pointers can continue to be `null`?

shiltian replied:
The idea is to change the semantics of `ptr null`. If we want to keep `ptr null` as a zero value, what would represent a real `nullptr`? I think last year we ruled out `ptr nullptr` because it's very confusing. I don't have a better idea.

shiltian commented on llvm/include/llvm/IR/DataLayout.h:
```diff
@@ -84,6 +84,11 @@ class DataLayout {
+    /// all-zeros, all-ones, or std::nullopt. Since we don't have a way to
+    /// represent an arbitrary bit pattern, we use std::nullopt to represent the
+    /// case where the nullptr value is neither 0 nor -1.
+    std::optional<APInt> NullPtrValue;
```
I'm also thinking that, to avoid a large number of test updates, maybe we should treat `std::nullopt` the same as the current behavior, which is a zero value. It might not make sense to update the data layout for all targets.

shiltian commented:
I also wonder if we want to introduce `ConstantZeroPointer` instead of using `ConstantExpr(inttoptr i8 0)`, since in long term it seems like we want to get rid of `ConstantExpr`. @nikic
